core/lib.ml

```diff
@@ -1693,10 +1695,25 @@ let env : (string * (located_primitive * Types.datatype * pure)) list = [
     (* CLI *)
     "getArgs",
     (`Server
-       (`PFun (fun _ _ ->
+       (pure_pfun (fun _ ->
             Value.(box_list (List.map box_string (Settings.get_rest_arguments ()))))),
      datatype "() ~> [String]",
+     IMPURE);
+
+    (* SPARQL *)
+    "sparql",
+    (`Server
+       (p3D (fun base uri query _ ->
```

Comment:
I think you can use p3 instead of p3D here.

Reply:
```suggestion
       (p3 (fun base uri query ->
```

Reply:
Nope, because p3 does not do Lwt.return.